lib/SIL/Utils/PrunedLiveness.cpp

```diff
@@ -442,32 +442,66 @@ template class PrunedLiveRange<MultiDefPrunedLiveness>;
 //                             SSAPrunedLiveness
 //===----------------------------------------------------------------------===//
 
-void SSAPrunedLiveness::findBoundariesInBlock(
-    SILBasicBlock *block, bool isLiveOut,
-    PrunedLivenessBoundary &boundary) const {
-  assert(isInitialized());
-
-  // For SSA, a live-out block cannot have a boundary.
-  if (isLiveOut)
-    return;
-
-  bool isDefBlockState = isDefBlock(block);
+/// Given live-within (non-live-out) \p block, find the last user.
+void findBoundaryInNonDefBlock(SILBasicBlock *block,
+                               PrunedLivenessBoundary &boundary,
+                               const PrunedLiveness &liveness) {
   for (SILInstruction &inst : llvm::reverse(*block)) {
-    if (isDefBlockState && isDef(&inst)) {
+    if (liveness.isInterestingUser(&inst)) {
+      boundary.lastUsers.push_back(&inst);
+      return;
+    }
+  }
+  llvm_unreachable("live-within block must contain an interesting use");
+}
+
+/// Given a live-within \p block that contains an SSA definition, and knowledge
+/// that all live uses are dominated by that single definition, find either the
+/// last user or a dead def.
+///
+/// A live range with a single definition cannot have any uses above that
+/// definition in the same block. This even holds for unreachable self-loops.
+void findBoundaryInSSADefBlock(SILNode *ssaDef,
+                               PrunedLivenessBoundary &boundary,
+                               const PrunedLiveness &liveness) {
+  // defInst is null for argument defs.
+  SILInstruction *defInst = dyn_cast<SILInstruction>(ssaDef);
+  for (SILInstruction &inst : llvm::reverse(*ssaDef->getParentBlock())) {
+    if (&inst == defInst) {
       boundary.deadDefs.push_back(cast<SILNode>(&inst));
       return;
     }
-    if (isInterestingUser(&inst)) {
+    if (liveness.isInterestingUser(&inst)) {
       boundary.lastUsers.push_back(&inst);
       return;
     }
   }
-  auto *deadArg = dyn_cast<SILArgument>(def);
-  assert(deadArg && deadArg->getParent() == block
+  auto *deadArg = dyn_cast<SILArgument>(ssaDef);
```

Comment:
Is this a `dyn_cast` rather than a `cast` in order to have a better assertion on the next line?

Reply:
Right, because `findBoundariesInBlock` is public, so it's easy for someone to use it wrong. And when the cast crashes I want them to realize it was their fault.